datafusion/physical-expr/src/aggregate/bit_and_or_xor.rs

```diff
@@ -0,0 +1,1103 @@
+use std::sync::Arc;
+
+use crate::{AggregateExpr, PhysicalExpr};
+use arrow::datatypes::DataType;
+use arrow::{
+    array::{
+        ArrayRef, Int16Array, Int32Array, Int64Array, Int8Array, UInt16Array,
+        UInt32Array, UInt64Array, UInt8Array,
+    },
+    datatypes::Field,
+};
+use datafusion_common::{downcast_value, DataFusionError, Result, ScalarValue};
+use datafusion_expr::Accumulator;
+
+use crate::aggregate::row_accumulator::{
+    is_row_accumulator_support_dtype, RowAccumulator,
+};
+use crate::aggregate::utils::down_cast_any_ref;
+use crate::expressions::format_state_name;
+use arrow::array::Array;
+use arrow::array::PrimitiveArray;
+use arrow::datatypes::ArrowNativeTypeOp;
+use arrow::datatypes::ArrowNumericType;
+use datafusion_row::accessor::RowAccessor;
+use std::ops::BitAnd as BitAndImplementation;
+use std::ops::BitOr as BitOrImplementation;
+use std::ops::BitXor as BitXorImplementation;
+
+fn bit_and<T>(array: &PrimitiveArray<T>) -> Option<T::Native>
+where
+    T: ArrowNumericType,
+    T::Native: BitAndImplementation<Output = T::Native> + ArrowNativeTypeOp,
+{
+    let null_count = array.null_count();
+
+    if null_count == array.len() {
+        return None;
+    }
+
+    let data: &[T::Native] = array.values();
+
+    match array.nulls() {
+        None => {
+            let bit_and = data
+                .iter()
+                .fold(T::Native::ONE.neg_wrapping(), |accumulator, value| {
+                    accumulator & *value
+                });
+
+            Some(bit_and)
+        }
+        Some(nulls) => {
+            let mut bit_and = T::Native::ONE.neg_wrapping();
+            let data_chunks = data.chunks_exact(64);
+            let remainder = data_chunks.remainder();
+
+            let bit_chunks = nulls.inner().bit_chunks();
+            data_chunks
+                .zip(bit_chunks.iter())
+                .for_each(|(chunk, mask)| {
+                    // index_mask has value 1 << i in the loop
+                    let mut index_mask = 1;
+                    chunk.iter().for_each(|value| {
+                        if (mask & index_mask) != 0 {
+                            bit_and = bit_and & *value;
+                        }
+                        index_mask <<= 1;
+                    });
+                });
+
+            let remainder_bits = bit_chunks.remainder_bits();
+
+            remainder.iter().enumerate().for_each(|(i, value)| {
+                if remainder_bits & (1 << i) != 0 {
+                    bit_and = bit_and & *value;
+                }
+            });
+
+            Some(bit_and)
+        }
+    }
+}
+
+fn bit_or<T>(array: &PrimitiveArray<T>) -> Option<T::Native>
+where
+    T: ArrowNumericType,
+    T::Native: BitOrImplementation<Output = T::Native> + ArrowNativeTypeOp,
+{
+    let null_count = array.null_count();
+
+    if null_count == array.len() {
+        return None;
+    }
+
+    let data: &[T::Native] = array.values();
+
+    match array.nulls() {
```

Comment:
This loop and the ones in `bit_and` are *very* similar -- (specifically the iteration over data chunks, the bit mask, and the remainder).  The only difference seems to be the initialization condition as well as how to update on each value.

Could you please try and refactor them to avoid the duplication?